mlx_lm/server.py

```diff
@@ -493,42 +493,96 @@ def generate_response(
         return response
 
     def get_prompt_cache(self, prompt):
+        """
+        Determines the portion of the prompt that needs processing by comparing
+        it to the cached prompt and attempting to reuse the common prefix.
+
+        This function updates the internal prompt cache state (tokens and model cache)
+        based on the comparison. If a common prefix exists, it attempts to trim
+        the model cache (if supported) to match the common prefix length, avoiding
+        recomputation.
+
+        Args:
+            prompt (List[int]): The tokenized new prompt.
+
+        Returns:
+            List[int]: The suffix of the prompt that actually needs to be processed
+                       by the model. This will be the full prompt if the cache is
+                       reset or cannot be effectively used.
+        """
         cache_len = len(self.prompt_cache.tokens)
         prompt_len = len(prompt)
-        prefix_len = min(cache_len, prompt_len)
+        com_prefix_len = common_prefix_len(self.prompt_cache.tokens, prompt)
+
+        # Default: process the full prompt
+        prompt_to_process = prompt
 
+        # Condition 1: Model changed or no common prefix at all. Reset cache.
         if (
             self.prompt_cache.model_key != self.model_provider.model_key
-            or prompt[:prefix_len] != self.prompt_cache.tokens[:prefix_len]
+            or com_prefix_len == 0
         ):
+            logging.debug(
+                f"*** Resetting cache (model change or no common prefix). com_prefix_len: {com_prefix_len} ***"
+            )
             self.prompt_cache.model_key = self.model_provider.model_key
             self.prompt_cache.cache = make_prompt_cache(self.model_provider.model)
             if self.model_provider.draft_model is not None:
                 self.prompt_cache.cache += make_prompt_cache(
                     self.model_provider.draft_model
                 )
```

Comment:
I think it's time to refactor this into a little helper to make it easier to change stuff in the future. Maybe make a function like `reset_prompt_cache`. It could even an inner function inside this one.